Non-interactive profiling mode

angstsmurf opened the pull request:
This is a quick-and-dirty implementation of #97. In a debug build, the game will look for a file named "autotesting.glkdata" which contains the name of an in-game test, run the test and, if it reaches the end of the game, print out the time it took.

It has some problems with the initial questions but seems to work. The autotesting file will need to have the standard header: "* //7B5A779B-4653-43DB-A516-F475DDC12987// autotesting" and then the name of the test on the next line, e.g. "me".

Note that some interpreters will not understand the ".glkdata" suffix, and instead look for a file named just "autotesting".

Any present Counterfeit Monkey-startup-data.glkdata file will break things, and should be deleted before each run.

I first tried to reuse the "testing.glkdata" file for this, but couldn't find a way to check whether a file is empty or not.

I would also be nice to be able to supply a list of actual commands rather than just the name of a pre-made internal test, but there seems to be plenty of pitfalls both when interpreting the external file and when feeding the commands to the interpreter.

curiousdannii commented:
I added a commit to use the low level command stream feature. It now just needs a file called autotesting.txt with a list of all the commands. What do you think?

angstsmurf commented:
Seems to work great! I wish I had known about the command stream earlier.

Game scripts that contain a restart, i.e. that answer RESTART to the final questions and then play through the game again, such as https://github.com/i7/counterfeit-monkey/blob/master/tools/command%20scripts/test_me-me2.txt seem to create an infinite loop, though.

curiousdannii commented:
Yeah the command stream is more like legacy code from I6 than a proper I7 feature. I think they even removed the verbs to use it!

Hmm restarts are tricky. Possibly it just needs to check if the stream already exists rather than forcibly recreating it.

angstsmurf commented:
I added a check for the command stream. We should keep track of total play time across restarts as well, so I guess we have to use another external file. Ideally, we should show time elapsed both since the start of the test and since last restart.

curiousdannii commented:
The rule runs after the recover objects rule, so all that's needed is checking if gg_commandstr is 0.

Hmm, we could call @protect on the `startTime` variable? I've never used that before, but this seems like a really simple case where it should work.

curiousdannii commented:
I updated Glk Object Recovery: https://github.com/i7/extensions/blob/master/Dannii%20Willis/Glk%20Object%20Recovery.i7x

angstsmurf commented:
Right, I added the updated Glk Object Recovery and implemented restart-resistant total test time, although I have a feeling it could be done in a cleaner way. 

Also, I'm not sure this will fail gracefully if the interpreter does not provide system clock functionality. Do you know of any such interpreter we could test it on? 

EDIT: I just noticed that this is broken on "old style" tests, when just a "testing" file is present. Then only the accumulated time across restarts will be shown. I'll see if I can fix it.

angstsmurf commented:
That should do it!

curiousdannii commented:
I don't have the time to properly test it sorry, but from what I can see it looks good to me!

Yeah, if we want to check the benefits from startup precomputation again we'll probably need a separate test.
